core/transaction-pool/graph/benches/basics.rs

```diff
@@ -0,0 +1,150 @@
+use criterion::{criterion_group, criterion_main, Criterion};
+
+use futures::executor::block_on;
+use substrate_transaction_graph::*;
+use sr_primitives::transaction_validity::ValidTransaction;
+use codec::Encode;
+use test_runtime::{Block, Extrinsic, Transfer, H256, AccountId};
+use sr_primitives::{
+	generic::BlockId,
+	transaction_validity::{TransactionValidity, TransactionTag as Tag},
+};
+use primitives::blake2_256;
+
+#[derive(Clone, Debug, Default)]
+struct TestApi {
+	nonce_limit: Option<u64>,
+}
+
+impl TestApi {
+	fn new_with_limit(limit: u64) -> Self {
+		TestApi { nonce_limit: Some(limit) }
+	}
+}
+
+fn to_tag(nonce: u64, from: AccountId) -> Tag {
+	let mut data = [0u8; 40];
+	data[..8].copy_from_slice(&nonce.to_le_bytes()[..]);
+	data[8..].copy_from_slice(&from.0[..]);
+	data.to_vec()
+}
+
+impl ChainApi for TestApi {
+	type Block = Block;
+	type Hash = H256;
+	type Error = error::Error;
+	type ValidationFuture = futures::future::Ready<error::Result<TransactionValidity>>;
+
+	fn validate_transaction(
+		&self,
+		_at: &BlockId<Self::Block>,
+		uxt: ExtrinsicFor<Self>,
+	) -> Self::ValidationFuture {
+		let nonce = uxt.transfer().nonce;
+		let from = uxt.transfer().from.clone();
+
+		let mut requires = vec![];
+		if nonce > 1 && self.nonce_limit.is_some() {
+			requires.push(to_tag(nonce-1, from.clone()));
+		};
+
+		let mut provides = vec![];
+		provides.push(to_tag(nonce, from.clone()));
+		if self.nonce_limit.is_some() {
+			provides.push(to_tag(nonce+1, from.clone()));
```

Comment:
That will definitely work, but the outcome will be really tricky. Basically if `nonce_limit` is set every transaction is actually providing two tags:
`nonce` and `nonce+1`
So this means that every two consecutive transactions will overlap.
Since the pool only supports exactly one transaction that provides given tag, when you first import `TxA(nonce, nonce+1)` and then `TxB (nonce+1, nonce+2)`, `TxA` will either be removed from the pool (if the priority of `TxB` is higher) or `TxB` will not be accepted. In the first case I think it may break the entire chain of transactions and move them to future since there will be no transaction that can provide `nonce`, so tags `required` by `TxB` will never be satisfied.